src/model.py

```diff
@@ -542,6 +552,30 @@ def verify_loras(self, request):
             verified_request = request
         return verified_request
 
+    def _check_health(self, requests):
+        coro = self.llm_engine.check_health()
+        future = asyncio.run_coroutine_threadsafe(coro, self._loop)
+        try:
+            future.result()
+        except Exception as e:
+            self.logger.log_error(
+                f"[vllm] Engine is not healthy and model will be unloaded: {e}"
+            )
+            pb_utils.unload_model(self.model_config["name"])  # non-blocking
```

Comment:
Technical question out of curiosity - not asking you to implement it here.

Can a python model reload itself? For example, would `pb_utils.load_model(self.model_config["name"])` work after `pb_utils.unload_model(self.model_config["name"])`?

It seems like some python logic within the model still runs after triggering the unload, and I'm curious what the expected behavior is here in general after a model calls unload on itself.

Reply:
I think it will work, as long as the BLS load is async, but a hack on the model file timestamp is needed to stop the core from running the update routine which will not reload anything.

Reply:
> It seems like some python logic within the model still runs after triggering the unload, and I'm curious what the expected behavior is here in general after a model calls unload on itself.

The unload is async, so it will begin immediately. The model lifecycle will unload / forget the model, but it only [releases](https://github.com/triton-inference-server/core/blob/main/src/model_repository_manager/model_lifecycle.h#L254) the [shared pointer](https://github.com/triton-inference-server/core/blob/main/src/model_repository_manager/model_lifecycle.h#L254) which is also [held by active inference requests](https://github.com/triton-inference-server/core/blob/main/src/infer_request.h#L798), so the model technically will not begin to unload until all active inference requests are completed.

Reply:
> so the model technically will not begin to unload until all active inference requests are completed.

How are they going to be completed if the engine is unhealthy?

Reply:
> How are they going to be completed if the engine is unhealthy?

This is an interesting point. To complete an inference request, all we need is to send the complete final flag. In other words, we don't need to pass the requests into the engine, but respond with an error and complete final flag as soon as they arrive at the model (i.e. those queued in the scheduler). We will also need a way to track the requests processing by the engine and respond to them as well.

Back to the original question, "can a python model reload itself?". The unload of the unhealthy model does not need to be completed before reloading and serving the reloaded the model. In other words, the unhealthy model and the reloaded model can coexist.